### PR TITLE
CUDA-graph-compatible releasing and resuming KV cache and model weight memory

### PR DESCRIPTION
**Pull request by fzyzcjy**
Related: https://github.com/sgl-project/sglang/issues/2542 and #2583

<details>
<summary>Outdated Content</summary>

The test will fail because it uses `LD_PRELOAD` currently (to intercept and change logic of cudaMalloc and cudaFree). If the general logic looks good, I will further update this PR to handle this part (e.g. try to specify LD_PRELOAD automatically when creating the backend process.)

## How to execute it

Suppose this branch of SGLang is at `/path/to/sglang`, then inside sglang's docker container, execute the following:

```
# install the torch_memory_saver (currently install from source, surely on pip later)
git clone https://github.com/fzyzcjy/torch_memory_saver
(cd torch_memory_saver && make reinstall)

cd /path/to/sglang
PYTHONPATH=$(pwd)/python LD_PRELOAD=/sgl-workspace/torch_memory_saver/torch_memory_saver_cpp.cpython-310-x86_64-linux-gnu.so python3 test/srt/test_release_gpu_occupation.py
```

</details>

Expected results are as follows. x is time, red color is memory consume. The low memory at the center is caused by temporarily release KV cache memory.

image

## What's changed

Though the PR seems large, most are boilerplate. 

Core: 

* Wrap things inside `with primary_memory_saver.region()`: TokenToKVPool.k_buffers/v_buffers, ModelRunner.model, ReqToTokenPool.req_to_token
* Call `primary_memory_saver.pause()/.resume()`: At `scheduler.py`, Scheduler.release_gpu_occupation/resume_gpu_occupation

Others:

* Add Engine.release_gpu_occupation/resume_gpu_occupation: Need to add multiple structs such as ReleaseGPUOccupationReqInput, and multiple forwarding boilerplate in Engine/TokenizerManager
* Add a base class BaseCausalLM: Need to change all models' parent classes.
* Add tests

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
(Well I only want to make it "not a draft" and not mean to @ so many people!)

**Comment by heiner** [comment]
Hey @fzyzcjy - very impressive!

QQ: Do you think the `LD_PRELOAD` could be avoided via something like this?

* KV caches use a special allocation (not `torch.empty` or anything else that goes via PyTorch) right away. That special allocation gives a pointer + size which PyTorch can use via e.g. `torch.frombuffer`.
* The special allocation has an explicit deletion + recreation with the `cuMemAddressReserve` + `cuMemMap` trick used in `torch_memory_saver`?

This way this could even be written in ctypes, or otherwise be independent of PyTorch itself.

**Comment by fzyzcjy** [comment]
@heiner Hi thank you for the suggestion!

Currently I do this because I also want to allow the model to be released as well, using one-liner `with memory_saver.region(): create_your_model_here`. 

Btw, I have updated how LD_PRELOAD is handled. Now users should not need to manually specify in command line.

(Also feel free to discuss in https://github.com/sgl-project/sglang/issues/2542, because I guess this comment will soon get folded once I push more commits...)

File: python/sglang/srt/models/base.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from torch import nn
+
+
+class BaseCausalLM(nn.Module):
```
Comment (merrymercy) [inline python/sglang/srt/models/base.py]: If we only need to add these two functions, we do not need to create a new base class here.

We can define two functions
 `def export_static_state(module)` and `def import_static_state(module, static_params)`

Just to minimize the big code change.

File: python/sglang/srt/managers/io_struct.py [hunk 2]
```diff
@@ -469,6 +469,26 @@ class GetWeightsByNameReqOutput:
     parameter: list
 
 
+@dataclass
+class ReleaseGPUOccupationReqInput:
```
Comment (merrymercy) [inline python/sglang/srt/managers/io_struct.py]: Call it `ReleaseMemoryOccupationReqInput`. Rename all places (code, api name, etc).

File: scripts/ci_install_dependency.sh [hunk 3]
```diff
@@ -14,6 +14,7 @@ pip install -e "python[all]" --find-links https://flashinfer.ai/whl/cu121/torch2
 
 # Force reinstall flashinfer
 pip install flashinfer==0.1.6 --find-links ${FLASHINFER_REPO} --force-reinstall --no-deps
+pip uninstall -y torch_memory_saver && pip install torch_memory_saver
```
Comment (merrymercy) [inline scripts/ci_install_dependency.sh]: use `--force-reinstall` instead of uninstall and reinstall?

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -792,6 +793,11 @@ def add_cli_args(parser: argparse.ArgumentParser):
             action="store_true",
             help="Delete the model checkpoint after loading the model.",
         )
+        parser.add_argument(
+            "--memory-saver",
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py]:
```suggestion
            "--enable-memory-saver",
```

File: python/sglang/srt/model_executor/model_runner.py [hunk 5]
```diff
@@ -536,6 +542,7 @@ def init_memory_pool(
             max_context_len=self.model_config.context_len + 4,
             device=self.device,
             use_records=False,
+            memory_saver_adapter=self.memory_saver_adapter,
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/model_runner.py]: Can we do something similar to how we handle `get_model`, so we do not need to pass `memory_saver_adapter` as an argument to all kinds of memory pools.
```
        with self.memory_saver_adapter.region():
            self.req_to_token_pool = ReqToTokenPool(
```

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/base.py (hunk 1 above)
Comment (fzyzcjy) [reply python/sglang/srt/models/base.py]: I was considering about customizability (e.g. a model may have different opinion about this), but yes this seems not needed at least today.

File: python/sglang/srt/models/llama.py [hunk 6]
```diff
@@ -329,7 +329,6 @@ def load_kv_cache_scales(self, quantization_param_path: str) -> None:
 
 
 class LlamaForCausalLM(nn.Module):
```
Comment (merrymercy) [inline python/sglang/srt/models/llama.py]: revert this?

File: python/sglang/srt/model_executor/model_runner.py [hunk 7]
```diff
@@ -590,6 +596,7 @@ def init_memory_pool(
             max_context_len=self.model_config.context_len + 4,
             device=self.device,
             use_records=False,
+            enable_memory_saver=self.server_args.enable_memory_saver,
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/model_runner.py:599]: See this comment https://github.com/sgl-project/sglang/pull/2630#discussion_r1912734551

**Review by merrymercy — COMMENTED** [review]
Almost there! Just one final comment.

File: python/sglang/srt/model_executor/model_runner.py (hunk 7 above)
Comment (fzyzcjy) [reply python/sglang/srt/model_executor/model_runner.py:599]: Oops I misunderstood your question. I am worried maybe not, because it can happen that we do have some tensors inside req pool in the future that needs to be preserved across memory release. (The torch_memory_saver does not offload things to cpu, instead it just throws away the content, in order to be faster)

~~But anyway, today it seems to be OK (will check it), so I can update it if needed.~~ Quickly skimmed it, worried maybe `BaseTokenToKVPool.free_slots` is one such tensor that hopes not to be released. Too tired now and not do any experiment though.

File: python/sglang/srt/models/llama.py (hunk 6 above)
Comment (fzyzcjy) [reply python/sglang/srt/models/llama.py]: done (CI seems to be flaky)
